cmd/habitat/node/node.go

```diff
@@ -0,0 +1,105 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+
+	"github.com/eagraf/habitat/cmd/habitat/api"
+	dataproxy "github.com/eagraf/habitat/cmd/habitat/data_proxy"
+	"github.com/eagraf/habitat/cmd/habitat/procs"
+	"github.com/eagraf/habitat/cmd/habitat/proxy"
+	"github.com/eagraf/habitat/cmd/sources"
+	"github.com/eagraf/habitat/pkg/compass"
+	"github.com/eagraf/habitat/pkg/p2p"
+	"github.com/eagraf/habitat/structs/ctl"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/viper"
+)
+
+const (
+	ReverseProxyHost = "0.0.0.0"
+	ReverseProxyPort = "2041"
+
+	P2PPort = "6000"
+)
+
+type Node struct {
+	ID string
+
+	ProcessManager *procs.Manager
+	P2PNode        *p2p.Node
+	ReverseProxy   *proxy.Server
+	DataProxy      *dataproxy.DataProxy
+}
+
+func NewNode() (*Node, error) {
+	priv, _ := compass.GetPeerIDKeyPair()
+
+	p2pNode, err := p2p.NewNode(P2PPort, priv)
+	if err != nil {
+		return nil, err
+	}
+	log.Info().Msgf("starting LibP2P node with peer ID %s listening at port %s", p2pNode.Host().ID().Pretty(), P2PPort)
+
+	procsDir := compass.ProcsPath()
+	reverseProxy := proxy.NewServer()
+
+	return &Node{
+		ID:             compass.NodeID(),
+		P2PNode:        p2pNode,
+		ReverseProxy:   reverseProxy,
+		DataProxy:      dataproxy.NewDataProxy(map[string]*sources.DataServerNode{}),
+		ProcessManager: procs.NewManager(procsDir, reverseProxy.Rules),
+	}, nil
+}
+
+func (n *Node) Start() error {
+	// Start reverse proxy
+	proxyAddr := fmt.Sprintf("%s:%s", ReverseProxyHost, ReverseProxyPort)
+	go n.ReverseProxy.Start(proxyAddr)
+
+	redirectURL, err := url.Parse("http://" + proxyAddr + "/habitat")
+	if err != nil {
+		log.Fatal().Err(err)
+	}
+	go proxy.LibP2PHTTPProxy(n.P2PNode.Host(), redirectURL)
+
+	// Start data proxy
+	sourcesPort := viper.GetString("data-proxy-port")
+	go n.DataProxy.Start(context.Background(), sourcesPort)
```

Comment:
ah this is making me really feel like we should abstract out at least reverse + data proxy into one

Reply:
Right - the more I think about it I think we underrated the reverse proxy as a really important part of the system. One idea I'm playing with is that we should only open two ports for all of Habitat: one for serving external API requests via HTTPS, and one for libp2p communication internally between community members. Applications can open other ports, but they are separate processes (we should probably still work out a port manager for applications)